scripts/model_gen_ng/model_gen_ng/templates/enum.tpl

```diff
@@ -0,0 +1,60 @@
+namespace [[ enum.namespace ]] {
+    [[ enum.doc | docstring ]]
+    enum Values
+    [%- if enum.underlying_type is defined %] : [[ enum.underlying_type ]][% endif %]{
+        Invalid,
+        [% for member in enum.members %]
+        [[- member.doc | docstring ]]
+        [[- member.name ]],
+        [% endfor %]
+        Count
+    };
+
+    inline llvm::StringRef getName(Values V) {
+        switch (V) {
+            case Invalid:
+                return "Invalid";
+            [%- for member in enum.members %]
+            case [[ member.name ]]:
+                return "[[ member.name ]]";
+            [%- endfor %]
+            default:
+                revng_abort();
+                break;
+        }
+    }
+
+    inline Values fromName(llvm::StringRef Name) {
+        [%- for member in enum.members %]
+        [%- if loop.first %]
+        if (Name == "[[ member.name ]]") {
+            return [[ member.name ]];
+        }
+        [%- else %]
+        else if (Name == "[[ member.name ]]") {
+            return [[ member.name ]];
+        }
+        [%- endif -%]
+        [%- endfor -%]
+        else {
+            return Invalid;
+        }
+    }
+
+    [[ enum.customcode ]]
+} // namespace [[ enum.namespace ]]
+
+template<>
+struct KeyedObjectTraits<[[ enum.fullname ]]>
+  : public IdentityKeyedObjectTraits<[[ enum.fullname ]]> {};
+
+namespace llvm::yaml {
+template<>
+struct ScalarEnumerationTraits<[[ enum.fullname ]]>
+  : public NamedEnumScalarTraits<[[ enum.fullname ]]> {
+  static constexpr const char Name[] = "[[ enum.name ]]";
+  [# TODO: do we want to include the Invalid member in the count or not? -#]
+  static constexpr const unsigned int NumMembers = [[ enum.members|length ]];
```

Comment:
Let's not abuse `llvm::yaml::*Traits`.
Make a new trait.

```c++
template<typename T>
constexpr size_t EnumElementsCount;

template<>
constexpr size_t EnumElementsCount<model::Architecture::Values> = 42;
```

Reply:
Premise: as of now, nobody is using `NumMembers`, I created it anticipating that sometime we might need it.
That said, I created a `revng/Model/EnumTraits.h` file with the template definition and instantiated the template for each enum, but isn't wouldn't it be more appropriate to add Name and NumMembers to our own NamedScalarEnumTraits?

Reply:
Remember that `NamedScalarEnumTraits` are used by the YAML I/O which does not consume these new fields. Not the best place to put them.